[MRG] Documentation updates for 4.0 release (#1283)
sourmash-bio/sourmash: doc/sourmash-sketch.md

Sourmash can work with unassembled reads; run
```
sourmash sketch dna -p k=21,k=31,k=51,abund metagenome.fq.gz
```
to compute three abundance-weighted sketches at k=21, 31, and 51, for the given FASTQ file.

### Protein sketches for genomes and proteomes

Likewise,
```
sourmash sketch translate genome.fna
```
will output a protein sketch in `./genome.fna.sig`, calculated by translating the genome sequence in all six frames and then using the default protein sketch parameters.

And
```
sourmash sketch protein -p k=25,scaled=500 -p k=27,scaled=250 genome.faa
```
outputs two protein sketches to `./genome.faa.sig`, one calculated with k=25 and scaled=500, the other calculated with k=27 and scaled=250.

If you want to use different encodings, you can specify them in a few ways; here is a parameter string that specifies a dayhoff encoding for the k-mers:
```
sourmash sketch protein -p k=25,scaled=500,dayhoff genome.faa
```

## More detailed documentation

### Input formats

`sourmash sketch` auto-detects and reads FASTQ or FASTA files, either uncompressed or compressed with gzip or bzip2. The filename doesn't matter; `sourmash sketch` will figure out the format from the file contents.

> You can also stream any of these formats into `sourmash sketch` via stdin by using `-` as the input filename.

Reviewer: Yes, that's true, but then the name of the sig is recorded as `-` which is really confusing when you compare a bunch of files. 

Also, should there be an example for how to do this?

### Input contents and output signatures

By default, `sourmash sketch` will produce signatures for each input *file*. If the file contains multiple FASTA/FASTQ records, these records will be merged into the output signature.

> 

Reviewer 2: Maybe we could add a little section on sketch naming? Would be nice to explain where names are used, default sketch naming, --name/--merge, --name-from-first, etc all in one spot. I really like how intuitive --name is, though it would definitely need a note on how using --name for a command with multiple files would perform a merge...

Author: ugh, I'm not sure even I can do this from memory... which makes it a pretty good idea 😆

Author: Added in dda99fec, please take a look! I don't know if your intent was to make me aware of how clunky and inflexible signature naming is in practice, but if so, it worked! - see https://github.com/dib-lab/sourmash/issues/1315 for some potential improvements :).

If you specify `--singleton`, `sourmash sketch` will produce signatures for each *record*.

> If you specify `--merge <name>`, sourmash sketch will produce signatures for all input files combined into one.

Reviewer: ```suggestion
If you specify `--merge <name>`, sourmash sketch will produce signatures for all input files and combine them into one signature.
```

This suggestion is my attempt to make it clear how `merge` is different than the default behavior. I'm not sure it's clear as the docs currently stand, but i'm also not sure that my suggestion fixes it.

Reviewer 2: Maybe we could add a little section on sketch naming? Would be nice to explain where names are used, default sketch naming, `--name`/`--merge`, `--name-from-first`, etc all in one spot. I really like how intuitive `--name` is, though it would definitely need a note on how using `--name` for a command with multiple files would perform a merge...

### Protein encodings

`sourmash sketch protein` and `sourmash sketch translate` output protein sketches by default, but can also use the `dayhoff` and `hp` encodings.  The [Dayhoff encoding](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-367/tables/1) collapses multiple amino acids into a smaller alphabet so that amino acids that share biochemical properties map to the same character. The hp encoding divides amino acids into hydrophobic and polar (hydrophilic) amino acids, collapsing amino acids with hydrophobic side chains together and doing the same for polar amino acids.

We are still in the process of benchmarking these encodings; ask [on the issue tracker](https://github.com/dib-lab/sourmash/issues) if you are interested in updates.

### Parameter strings

The `-p` argument to `sourmash sketch` provides parameter strings to sourmash, and these control what signatures and sketches are calculated and output. Zero or more parameter strings can be given to sourmash. Each parameter string produces at least one sketch.

A parameter string is a space-delimited collection that can contain one or more fields, comma-separated.

> * `k=<ksize>` - compute a sketch at this k-mer size; can provide more than one time in a parameter string. Typically `ksize` is between 4 and 100.

Reviewer: ```suggestion
* `k=<ksize>` - create a sketch at this k-mer size; can provide more than one time in a parameter string. Typically `ksize` is between 4 and 100.
```

### More complex parameter string examples

Below are some more complicated `sourmash sketch` command lines:

* `sourmash sketch dna -p k=51` - default to a scaled=1000 and noabund for a k-mer size of 51 (based on moltype/command)
* `sourmash sketch dna -p k=31,k=51,k=21` - compute multiple ksizes, using the defaults otherwise
* `sourmash sketch translate -p k=20,num=500,protein -p k=19,num=400,dayhoff,abund -p k=30,scaled=200,hp` - compute multiple ksizes, moltypes, and scaled/num.

### Locations for output files

Signature files can contain multiple signatures and sketches. Use `sourmash sig describe` to get details on the contents of a file.

You can use `-o <filename>` to specify a file output location for all the output signatures; `-o -` means stdout. This does not merge signatures unless `--merge` is provided.

Specify `--outdir` to put all the signatures in a specific directory.

### Downsampling and flattening signatures

> Calculating signatures is probably the most time consuming part of using sourmash, and it is the only part that requires access to the raw data. Moreover, the output signatures are generally much smaller than the input data. So, we generally suggest calculating a large set of signatures once.

Reviewer: ```suggestion
Creating signatures is probably the most time consuming part of using sourmash, and it is the only part that requires access to the raw data. Moreover, the output signatures are generally much smaller than the input data. So, we generally suggest creating a large set of signatures once.
```

To support this, sourmash can do two kinds of signature conversion without going back to the raw data.

> First, you can downsample `num` and `scaled` signatures using `sourmash sig downsample`.  For any sketch calculated with `num` parameter, you can decrease that `num`. And, for any `scaled` parameter, you can increase the `scaled`. This will decrease the size of the sketch accordingly; for example, going from a num of 5000 to a num of 1000 will decrease the sketch size by a factor of 5, and going from a scaled of 1000 to a scaled of 10000 will decrease the sketch size by a factor of 10.

Reviewer: ```suggestion
First, you can downsample `num` and `scaled` signatures using `sourmash sig downsample`.  For any sketch created with `num` parameter, you can decrease that `num`. And, for any `scaled` parameter, you can increase the `scaled`. This will decrease the size of the sketch accordingly; for example, going from a `num` of 5000 to a `num` of 1000 will decrease the sketch size by a factor of 5, and going from a `scaled` of 1000 to a `scaled` of 10000 will decrease the sketch size by a factor of 10.
```

> (Note that decreasing num or increasing scaled will increase calculation speed and lower the accuracy of your results.)

Reviewer: ```suggestion
(Note that decreasing `num` or increasing `scaled` will increase calculation speed and lower the accuracy of your results.)
```

> Second, you can flatten abundances using `sourmash sig flatten`. For any sketch calculated with `abund`, you can convert it to a `noabund` sketch.  This will  decrease the sketch size, although not necessarily by a lot.

Reviewer: ```suggestion
Second, you can flatten abundances using `sourmash sig flatten`. For any sketch created with `abund`, you can convert it to a `noabund` sketch.  This will decrease the sketch size, although not necessarily by a lot.
```

> Unfortunately, changing the k-mer size or using different DNA/protein encodings cannot be done on a sketch, and you need to calculate new signatures from the raw data for that.

Reviewer: ```suggestion
Unfortunately, changing the k-mer size or using different DNA/protein encodings cannot be done on a sketch, and you need to create new signatures from the raw data for that.
```
